### PR TITLE
Drop implicit H that gets added for negative N and other fixes

### PR DESCRIPTION
**Pull request by ChayaSt**
This PR fixes several issues:

- [x] Remove extraneous checks when checking for molecule instance (#19)

- [x] Better handling of connectivity table bond order

- [x] Do not perceive bond order when loading an oemol from a JSON QCArchive molecule (#21)

- [x] Generate rdmol with explicit hydrogen is SMILES has explicit hydrogen 

- [x] When explicit H has map indices, rdkit will use that information when checking for potential sterocenters #20. This PR adds a workaround. 

### Testing 
The laster version of rdkit (2019.03.1) has better handling of stereoisomers and also now considers a N in a three membered ring as stereogenic. (https://github.com/rdkit/rdkit/issues/2268)
This broke some tests for canonical isomeric SMILES and InChIs. 
The older version of rdkit dropped some stereo information for these SMILES:
```
'C[N@@+]12C[C@@H]([C@@H](CC1)CC2)OC(=O)C(O)(c3ccccc3)c4ccccc4',
'C[C@@]12CC[C@@H](CC1)C(C)(C)O2',
'[H][C@]12C(=O)N(C(=O)[C@@]1([H])[C@@]1(CC[C@@]2([H])CC1)NC(=O)OCC(O)=O)C1=CC=C(NC(C)=O)C=C1',
'C[C@@]12OO[C@@](CCC(O)=O)(C3=CC=CC=C13)C1=C2C=CC=C1',
'CC(N1CCN(CC1)c2ncc(cc2)C(F)(F)F)(C(=O)N[C@H]3[C@@H]4C[C@@]5(CC(C4)C[C@H]3C5)C(=O)N)C',
'CC1(C)[C@H]2CC[C@]1(C)CC2'
```
3 SMILES that have a N in a three membered ring now behave differently between rdkit and openeye.

1. `CC1=C(C(=O)C2=C(C1=O)[N@@]3C[C@H]4[C@@H]([C@@]3([C@@H]2COC(=O)N)OC)N4)N Mitomycin` - The N here has an H bonded to it. OE does not consider it stereogenic (https://github.com/openforcefield/openforcefield/issues/146#issuecomment-490177618)
2. `C1[C@H]2N1C(=O)N=C2N Imexon` - Corner case. The N is bridged with the C and therefore the chirality of the N is implicitly defined. RDKit considers it undefined but openeye does not
3. `[H][C@@]1(C)C[N@@]1C1=CC(=O)C2=C(C(CO)=C(N2C)C2=CC=CC=C2)C1=O` - This might be a bug in rdkit. It seems like rdkit ignores the stereo information on the N and considers the chirality undefined. 

Because of these changes and fixes to the tests, CI environments were changed to start testing against rdkit 2019 and later. 
The 3 SMILES that behave differently between rdkit and openey were added to a separate test that only tests it against openeye.

### CODE REVIEW

**Comment by ChayaSt** [comment]
The following fixes were added to this PR:
1. Handle non integer bond orders in connectivity table
2. Fixed provenance
3. Do not remove explicit hydrogen when loading an rdkit molecule from an explicit H SMILES (This is not the default behavior in rdkit)
4. This fix introduced spurious stereo warnings because rdkit considers the the map indices on H when checking for stereo. Removing the map indices and storing them in atom data does not help because that information is used as well.

File: cmiles/_cmiles_oe.py [hunk 1]
```diff
@@ -33,7 +33,10 @@ def mol_from_json(symbols, connectivity, geometry, permute_xyz=False):
     for bond in connectivity:
         a1 = molecule.GetAtom(oechem.OEHasAtomIdx(bond[0]))
         a2 = molecule.GetAtom(oechem.OEHasAtomIdx(bond[1]))
-        molecule.NewBond(a1, a2, bond[-1])
+        bond_order = bond[-1]
+        if not isinstance(bond_order, int) and not bond_order.is_integer():
```
Comment (j-wags) [inline cmiles/_cmiles_oe.py]: Good thing to check. I had just run into un-`kekulize`d molecule problems last week.

File: cmiles/_cmiles_rd.py [hunk 2]
```diff
@@ -274,6 +274,7 @@ def get_symbols(molecule):
 
 
 def has_stereo_defined(molecule):
+    #ToDo Fix this function for carbons with explicit hydrogen and imine with implicit hydrogens
```
Comment (j-wags) [inline cmiles/_cmiles_rd.py]: Thanks for keeping records of these thoughts. It'll really help moving forward!

File: cmiles/tests/reference/smiles_rd_18_19_change.smi [hunk 3]
```diff
@@ -0,0 +1,4 @@
+CC1=C(C(=O)C2=C(C1=O)[N@@]3C[C@H]4[C@@H]([C@@]3([C@@H]2COC(=O)N)OC)N4)N Mitomycin N in 3 membered ring has H which oe does not flag as stereogenic
```
Comment (j-wags) [inline cmiles/tests/reference/smiles_rd_18_19_change.smi]: 👍

File: cmiles/tests/reference/generate_reference.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+def get_smiles_list(f):
+    smiles = open(f, 'r').read().split('\n')[:-1]
+    names = []
+    for i in range(len(smiles)):
+        smiles_name = smiles[i].split(' ')
+        smiles[i] = smiles_name[0]
+        names.append(smiles_name[-1])
+    return smiles, names
+#smiles, name = get_smiles_list('drug_bank_stereo.smi')
```
Comment (j-wags) [inline cmiles/tests/reference/generate_reference.py]: Is it safe to clean up unused code in this file (like `get_smiles_list`)? And adding a small docstring covering the purpose of this file would be helpful for when I come back to this after you're gone.

File: cmiles/tests/test_cmiles.py [hunk 5]
```diff
@@ -727,4 +733,9 @@ def test_permute_xyz(toolkit):
     permuted_geom = np.asarray(permuted_hooh['geometry']).reshape(int(len(hooh['geometry'])/3), 3)
     for m in atom_map:
         for i in range(3):
-            assert json_geom[atom_map[m]][i] == pytest.approx(permuted_geom[m-1][i], 0.0000001)
\ No newline at end of file
+            assert json_geom[atom_map[m]][i] == pytest.approx(permuted_geom[m-1][i], 0.0000001)
+
+@pytest.mark.parametrize('toolkit', toolkits_name)
+def test_provenance(toolkit):
```
Comment (j-wags) [inline cmiles/tests/test_cmiles.py]: Sweet! Loving this provenance.

File: cmiles/tests/test_utils.py [hunk 6]
```diff
@@ -162,33 +162,15 @@ def test_canonical_order_rd():
 @pytest.mark.parametrize('input, output', [('COC(C)c1c(Cl)ccc(F)c1Cl', False),
                                            ('C[C@@H](c1c(ccc(c1Cl)F)Cl)OC', False),
                                            ('O=O', True),
-                                           ('[CH3:1][CH2:3][CH2:4][CH3:2]', False)])
+                                           ('[CH3:1][CH2:3][CH2:4][CH3:2]', False),
```
Comment (j-wags) [inline cmiles/tests/test_utils.py]: You are a master of unit tests and reproducibility, high on a mountaintop above us mere mortals.

File: cmiles/tests/test_utils.py [hunk 7]
```diff
@@ -430,3 +412,72 @@ def test_atom_order_in_mol_copy(toolkit, smiles):
...
+    }
+    mol = utils.load_molecule(hooh)
+    assert isinstance(mol, oechem.OEMol)
+
+    hooh['connectivity'] = [[0, 1, 1.0], [1, 2, 1.0], [2, 3, 1]]
+    mol = utils.load_molecule(hooh)
+    assert isinstance(mol, oechem.OEMol)
+
+    hooh['connectivity'] = [[0, 1, 1.0], [1, 2, 1.5], [2, 3, 1.0]]
+    with pytest.raises(ValueError):
```
Comment (j-wags) [inline cmiles/tests/test_utils.py]: Nice

File: cmiles/tests/reference/generate_reference.py (hunk 4 above)
Comment (ChayaSt) [reply cmiles/tests/reference/generate_reference.py]: I removed the unused code. It's not needed and unimportant.

**Pull request merged**